### PR TITLE
Port DNA (getting started) example to C++

### LINKED ISSUE #3752: Port "getting started" example code to C++
also known as the DNA example

### PR DESCRIPTION
**Pull request by Wumpf**
### What

* Fixes #3752
* docs not included yet #3753

Also adds implicit conversion to underlying datatypes. Seems quite convenient, to be seen if this can get in the way.

image

### Checklist
* [x] I have read and agree to [Contributor Guide](https://github.com/rerun-io/rerun/blob/main/CONTRIBUTING.md) and the [Code of Conduct](https://github.com/rerun-io/rerun/blob/main/CODE_OF_CONDUCT.md)
* [x] I've included a screenshot or gif (if applicable)
* [x] I have tested [demo.rerun.io](https://demo.rerun.io/pr/3963) (if applicable)
* [x] The PR title and labels are set such as to maximize their usefulness for the next release's CHANGELOG

- [PR Build Summary](https://build.rerun.io/pr/3963)
- [Docs preview](https://rerun.io/preview/2deaabee322849c5fba601b83f2b0194fd72b68a/docs) 
- [Examples preview](https://rerun.io/preview/2deaabee322849c5fba601b83f2b0194fd72b68a/examples) 
- [Recent benchmark results](https://ref.rerun.io/dev/bench/)
- [Wasm size tracking](https://ref.rerun.io/dev/sizes/)

### CODE REVIEW

File: examples/cpp/dna/main.cpp [hunk 1]
```diff
@@ -0,0 +1,78 @@
+#include <rerun.hpp>
+#include <rerun/demo_utils.hpp>
+
+#include <random>
+
+using namespace rerun::demo;
+
+static constexpr size_t NUM_POINTS = 100;
+
+int main(int argc, char** argv) {
```
Comment (emilk) [inline examples/cpp/dna/main.cpp:10]: unused args

```suggestion
int main() {
```

File: examples/cpp/dna/README.md [hunk 2]
```diff
@@ -0,0 +1,29 @@
+---
+title: Helix
```
Comment (emilk) [inline examples/cpp/dna/README.md:2]: not new in this PR, but it's a bit confusing that the example is called both `dna` and `Helix`

File: rerun_cpp/src/rerun/components/line_strip3d_ext.cpp [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+
+// #define EDIT_EXTENSION
+
+namespace rerun {
+    namespace components {
+
+#ifdef EDIT_EXTENSION
+        // [CODEGEN COPY TO HEADER START]
+
+        /// Create linestrip from a list of vectors.
```
Comment (emilk) [inline rerun_cpp/src/rerun/components/line_strip3d_ext.cpp:11]:
```suggestion
        /// Create line strip from a list of positions, each connected to the next.
```

File: rerun_cpp/src/rerun/demo_utils.cpp [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+            std::vector<components::Color>& out_colors
+        ) {
+            out_points.reserve(num_points);
+            out_colors.reserve(num_points);
+
+            for (size_t i = 0; i < num_points; ++i) {
+                float angle = static_cast<float>(i) * angular_step * TAU + angular_offset;
+                out_points.push_back({
+                    sinf(angle) * radius,
+                    cosf(angle) * radius,
```
Comment (emilk) [inline rerun_cpp/src/rerun/demo_utils.cpp:33]: cos,sin is the x,y order

```suggestion
                    cosf(angle) * radius,
                    sinf(angle) * radius,
```

File: rerun_cpp/src/rerun/demo_utils.cpp [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+                    static_cast<float>(i) * z_step,
+                });
+
+                out_colors.push_back(colormap_turbo_srgb(static_cast<float>(i) / num_points));
+            }
+        }
+
+        template <size_t N>
+        static float dot(const float (&a)[N], const float (&b)[N]) {
+            float sum = 0.0f;
```
Comment (emilk) [inline rerun_cpp/src/rerun/demo_utils.cpp:43]: nitpick, but it is always a good idea to use higher precision accumulators

```suggestion
            double sum = 0.0f;
```

**Review by emilk — APPROVED** [review]

File: rerun_cpp/src/rerun/demo_utils.cpp (hunk 4 above)
Comment (Wumpf) [reply rerun_cpp/src/rerun/demo_utils.cpp:33]: will fix in all languages then 🤷  (got this from Rust)

File: examples/cpp/dna/README.md (hunk 2 above)
Comment (Wumpf) [reply examples/cpp/dna/README.md:2]: helix is the better name imho. DNA sounds like we're doing actual medical visualization

**Comment by Wumpf** [comment]
known error on cargo deny.

**Pull request merged**